topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,238 @@
+		logging.GetLogger().Debugf(
+			"Libvirt interface %s on %s", name, itf.Host)
+		node := probe.graph.LookupFirstNode(
+			graph.Metadata{"Name": name, "Type": "tun"})
+		if node != nil {
+			probe.EnrichNode(node, itf)
+		}
+	}
+}
+
+// EnrichNode adds the libvirt interface information to a node of the graph
+func (probe *Probe) EnrichNode(node *graph.Node, itf *Interface) {
+	tr := probe.graph.StartMetadataTransaction(node)
+	tr.AddMetadata("Libvirt.MAC", itf.Mac.Address)
+	tr.AddMetadata("Libvirt.Domain", itf.Host)
+	address := itf.Address
+	formatted := fmt.Sprintf(
+		"%s:%s.%s.%s.%s", address.Type, address.Domain, address.Bus,
+		address.Slot, address.Function)
+	tr.AddMetadata("Libvirt.Address", formatted)
+	tr.AddMetadata("PeerIntfMAC", itf.Mac.Address)
+	tr.Commit()
+}
+
+// Start get all domains attached to a libvirt connection
+func (probe *Probe) Start() {
+	// The event loop must be registered WITH its poll loop active BEFORE the
+	// connection is opened. Otherwise it just does not work.
+	err := libvirtgo.EventRegisterDefaultImpl()
+	if err != nil {
+		logging.GetLogger().Errorf("libvirt event handler:  %s", err)
+	}
+
+	go func() {
+		for {
+			err = libvirtgo.EventRunDefaultImpl()
+			if err != nil {
+				logging.GetLogger().Errorf("libvirt poll loop problem: %s", err)
+			}
+		}
+	}()
+
+	conn, err := libvirtgo.NewConnectReadOnly(probe.uri)
+	if err != nil {
+		logging.GetLogger().Errorf("Failed to create libvirt connect")
+		return
+	}
+	probe.conn = conn
+	callback := func(
+		c *libvirtgo.Connect, d *libvirtgo.Domain,
+		event *libvirtgo.DomainEventLifecycle,
+	) {
+		if event.Event == libvirtgo.DOMAIN_EVENT_STARTED {
+			var interfaces []*Interface
+			probe.QueryDomain(d, &interfaces)
+			probe.RegisterInterfaces(interfaces)
+		}
+	}
+	callbackID, err := conn.DomainEventLifecycleRegister(nil, callback)
+	probe.callbackID = callbackID
+	if err != nil {
+		logging.GetLogger().Errorf(
+			"Could not register the lifecycle event handler %s", err)
+	}
+	domains, err := probe.conn.ListAllDomains(0)
+	if err != nil {
+		logging.GetLogger().Error(err)
+		return
+	}
+	var interfaces []*Interface
+	for _, domain := range domains {
+		probe.QueryDomain(&domain, &interfaces)
+	}
+	probe.RegisterInterfaces(interfaces)
+}
+
+// Stop stops the probe
+func (probe *Probe) Stop() {
+	if probe.callbackID != -1 {
+		probe.conn.DomainEventDeregister(probe.callbackID)
+	}
+	probe.conn.Close()
+}
+
+// NewProbe creates a libvirt topology probe
+func NewProbe(g *graph.Graph, uri string) *Probe {
+	probe := &Probe{
+		graph:        g,
+		interfaceMap: make(map[string]*Interface),
+		callbackID:   -1,
+		uri:          uri,
+	}
+	g.AddEventListener(probe)
+	return probe
+}
+
+// NewProbeFromConfig initializes the probe
+func NewProbeFromConfig(g *graph.Graph) (*Probe, error) {
+	uri := config.GetString("libvirt.url")
+	probe := NewProbe(g, uri)
+	return probe, nil
+}
+
+// OnNodeUpdated event
+func (probe *Probe) OnNodeUpdated(n *graph.Node) {
```

Comment:
Not sure we should handle the case where the interface is renamed

Reply:
I don't think you can rename libvirt interface like this. But the code did not handle the dynamic addition of interfaces and this has been corrected with a new callback.